### PR TITLE
New: Add TorrentRssIndexer

### LINKED ISSUE #248: Support for Torrent RSS Feeds
**Is your feature request related to a problem? Please describe.**
I have Torrent RSS Feeds in Sonarr/Radarr, but I cannot add them to Prowlarr.

**Describe the solution you'd like**
An Indexer Type in Prowlarr matching the functionality of Torrent RSS Feeds from Sonarr/Radarr

**Describe alternatives you've considered**
My current alternative is leaving the Torrent RSS Feed configured manually in each *arr app.

**Additional context**
I have one Indexer that wishes to remain REALLY anonymous. The forum software does not use TorzNab format, but it does allow for RSS feeds. Before Prowlarr, I have configured this site into *arr apps as a Torrent RSS Feed and it works wonderfully. I would like to add it to Prowlarr simply so it can sync across all my apps with all my other Indexers.

### LINKED ISSUE #1101: Ability to add generic RSS
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe

Currently only way to add another site is to utilize Generic Tornznab, but not all sites support that. Would be great to just have the ability to add a generic rss feed just like sonarr/radarr Torrent RSS Feed option.

### Describe the solution you'd like

Adding a Generic Torrent RSS Feed modeled after sonarr/radarr

### Describe alternatives you've considered

can do individually in the arr itself, but thats the whole purpose of prowlarr, to centralize that. Also very handy for cross-seed.

### Anything else?

no

[AB#3477](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/3477)

### PR DESCRIPTION
**Pull request by mynameisbogdan**
#### Database Migration
NO

#### Description
`Torrent RSS Feed` was requested.

* Fixes #1101
* Fixes #248

### CODE REVIEW

**Comment by bakerboy448** [comment]
Towards
- https://github.com/Prowlarr/Prowlarr/issues/1101
- https://github.com/Prowlarr/Prowlarr/issues/248

File: src/NzbDrone.Core/Indexers/Definitions/TorrentRss/TorrentRssIndexer.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace NzbDrone.Core.Indexers.Definitions.TorrentRss
+{
+    public class TorrentRssIndexer : TorrentIndexerBase<TorrentRssIndexerSettings>
+    {
+        private readonly ITorrentRssParserFactory _torrentRssParserFactory;
+
+        public override string Name => "Torrent RSS Feed";
+        public override string[] IndexerUrls => new[] { "" };
+        public override string Description => "";
```
[the remark this answers is not in the document]
Comment (bakerboy448) [reply src/NzbDrone.Core/Indexers/Definitions/TorrentRss/TorrentRssIndexer.cs:14]:
```suggestion
        public override string Description => "Generic RSS Feed containing torrents. Requires X Y Z RSS attributes";
```

or see faq entry for attributes

where X Y Z are the min required rss attributes

[the remark this answers is not in the document]
Comment (bakerboy448) [reply src/NzbDrone.Core/Indexers/Definitions/TorrentRss/TorrentRssIndexer.cs]: we'll want to add the FAQ entry name as well whatever the header is named.

File: src/NzbDrone.Core/Indexers/Definitions/TorrentRss/TorrentRssIndexerSettings.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        [FieldDefinition(0, Label = "Full RSS Feed URL")]
+        public string BaseUrl { get; set; }
+
+        [FieldDefinition(1, Label = "Cookie", HelpText = "If you site requires a login cookie to access the rss, you'll have to retrieve it via a browser.")]
+        public string Cookie { get; set; }
+
+        [FieldDefinition(2, Type = FieldType.Checkbox, Label = "Allow Zero Size", HelpText="Enabling this will allow you to use feeds that don't specify release size, but be careful, size related checks will not be performed.")]
+        public bool AllowZeroSize { get; set; }
+
+        [FieldDefinition(2, Type = FieldType.Number, Label = "Default Size", HelpText="Add a default size for feeds with missing sizes.", Unit = "MB", Advanced = true)]
```
Comment (Qstick) [inline src/NzbDrone.Core/Indexers/Definitions/TorrentRss/TorrentRssIndexerSettings.cs:37]:
```suggestion
        [FieldDefinition(3, Type = FieldType.Number, Label = "Default Size", HelpText="Add a default size for feeds with missing sizes.", Unit = "MB", Advanced = true)]
```

File: src/NzbDrone.Core/Indexers/Definitions/TorrentRss/TorrentRssSettingsDetector.cs [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+        {
+            _httpClient = httpClient;
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Detect settings for Parser, based on URL
+        /// </summary>
+        /// <param name="settings">Indexer Settings to use for Parser</param>
+        /// <returns>Parsed Settings or <c>null</c></returns>
```
Comment (Qstick) [inline src/NzbDrone.Core/Indexers/Definitions/TorrentRss/TorrentRssSettingsDetector.cs:37]: Purge this as we aren't maintaining summary tags as a code style choice

File: src/NzbDrone.Core/Indexers/Definitions/TorrentRss/TorrentRssIndexer.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bakerboy448) [reply src/NzbDrone.Core/Indexers/Definitions/TorrentRss/TorrentRssIndexer.cs]: https://wiki.servarr.com/en/prowlarr/faq#can-i-add-any-generic-torrent-rss-feed

needs to be populated with the required attributes.

File: src/NzbDrone.Core/Indexers/Definitions/TorrentRss/TorrentRssSettingsDetector.cs [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+        private void ValidateReleaseSize(TorrentInfo[] releases, TorrentRssIndexerSettings indexerSettings)
+        {
+            if (!indexerSettings.AllowZeroSize && releases.Any(r => r.Size == 0))
+            {
+                throw new UnsupportedFeedException("Feed doesn't contain the release content size.");
+            }
+
+            if (releases.Any(r => r.Size != 0 && r.Size < ValidSizeThreshold))
+            {
+                throw new UnsupportedFeedException("Size of one more releases lower than {0}, feed must contain release content size.", ValidSizeThreshold.SizeSuffix());
```
Comment (ilike2burnthing) [inline src/NzbDrone.Core/Indexers/Definitions/TorrentRss/TorrentRssSettingsDetector.cs:291]:
```suggestion
                throw new UnsupportedFeedException("Size of one or more releases lower than {0}, feed must contain release content size.", ValidSizeThreshold.SizeSuffix());
```

File: src/NzbDrone.Core/Indexers/Definitions/TorrentRss/TorrentRssIndexerSettings.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        public TorrentRssIndexerSettings()
+        {
+            BaseUrl = string.Empty;
+            AllowZeroSize = false;
+        }
+
+        [FieldDefinition(0, Label = "Full RSS Feed URL", HelpLink = "https://wiki.servarr.com/en/prowlarr/faq#can-i-add-any-generic-torrent-rss-feed")]
+        public string BaseUrl { get; set; }
+
+        [FieldDefinition(1, Label = "Cookie", HelpText = "If you site requires a login cookie to access the rss, you'll have to retrieve it via a browser.")]
```
Comment (ilike2burnthing) [inline src/NzbDrone.Core/Indexers/Definitions/TorrentRss/TorrentRssIndexerSettings.cs:31]:
```suggestion
        [FieldDefinition(1, Label = "Cookie", HelpText = "If your site requires a login cookie to access the RSS, you'll have to retrieve it via a browser.")]
```

**Review by ilike2burnthing — CHANGES_REQUESTED** [review]
Back again with the vital suggested changes ;)

**Comment by bakerboy448** [comment]
Oooh that's beautiful 😭

**Review by Qstick — APPROVED** [review]

**Pull request merged**